### PR TITLE
Add DuckDNS DNS API

### PR DESCRIPTION
**Pull request by RockyTV**
https://www.duckdns.org/spec.jsp

### CODE REVIEW

File: dnsapi/dns_duckdns.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/bash
```
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: please use `sh` shebang instead

File: dnsapi/dns_duckdns.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  _debug "txtvalue: $txtvalue"
+
+  if [ -z $DUCKDNS_TOKEN ]; then
+    _err "DuckDNS token is not defined."
+    _err "Please export your account token to DUCKDNS_TOKEN and try again."
+    return 1
+  fi
+
+  duckdns_url="https://www.duckdns.org/update?domains=$_domain&token=$DUCKDNS_TOKEN&clear=true"
+  if [[ $(_get $duckdns_url) -ne "OK" ]]; then
```
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: do not use double bracket

File: dnsapi/dns_duckdns.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+dns_duckdns_add() {
+  fulldomain=$1
+  txtvalue=$2
+
+  _info "Adding TXT record to DuckDNS domain"
+  _debug "fulldomain: $fulldomain"
+  _debug "txtvalue: $txtvalue"
+  
+  if [ -z "$DUCKDNS_TOKEN" ]; then
+    _err "DuckDNS token is not defined."
```
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: add `DUCKDNS_TOKEN=""`  here.

File: dnsapi/dns_duckdns.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    _err "Please export your account token to DUCKDNS_TOKEN and try again."
+    return 1
+  fi
+  _saveaccountconf DUCKDNS_TOKEN "$DUCKDNS_TOKEN"
+  
+  duckdns_url="https://www.duckdns.org/update?domains=$_domain&token=$DUCKDNS_TOKEN&txt=$txtvalue"
```
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: where  does `$_domain` come from ?

Comment (RockyTV) [inline dnsapi/dns_duckdns.sh]: It comes from acme.sh itself. A lot of DNS APIs use it, take a look: https://github.com/Neilpang/acme.sh/search?utf8=%E2%9C%93&q=_domain&type=

Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: you should only use `$fulldomain`,  and detect which part is the root zone, and which part is the  sub domain zone.   do not rely on other variables. in most of the other dns api hooks, there is a function `getroot()`.  please take a look.

**Comment by RockyTV** [comment]
Should be good to go now.

**Comment by raidenii** [comment]
The API URL used in removing the TXT record is wrong, you'll need to include txt in the URI as well, otherwise it will remove the A and AAAA record.
I have forked my version: e7dff47

**Comment by Neilpang** [comment]
duckdns   is already supported,  so close this.

**Pull request closed without merging**